django/contrib/gis/db/backends/postgis/schema.py

```diff
@@ -36,3 +36,29 @@ def _create_index_sql(self, model, fields, suffix="", sql=None):
             "columns": field_column,
             "extra": '',
         }
+
+    def _alter_column_type_sql(self, table, old_field, new_field, new_type):
+        """
+        Special case when dimension changed.
+        """
+        if not hasattr(old_field, 'dim') or not hasattr(new_field, 'dim'):
+            return super(PostGISSchemaEditor, self)._alter_column_type_sql(
+                table, old_field, new_field, new_type
+            )
+
+        if old_field.dim == 2 and new_field.dim == 3:
+            sql_alter = "ALTER COLUMN %(column)s TYPE %(type)s USING ST_Force3D(%(column)s)::%(type)s"
+        elif old_field.dim == 3 and new_field.dim == 2:
+            sql_alter = "ALTER COLUMN %(column)s TYPE %(type)s USING ST_Force2D(%(column)s)::%(type)s"
+        else:
+            sql_alter = self.sql_alter_column_type
+        return (
```

Comment:
This looks a bit odd to me. I wonder if something like this could be a bit more readable? Up to you.
```
return (
    (sql_alter % {"column": self.quote_name(new_field.column), "type": new_type}, []),
    [],
)
```